README.md
# Random color generator

[![pub package](https://img.shields.io/pub/v/random_color.svg)](https://pub.dartlang.org/packages/random_color) 
[![CodeFactor](https://www.codefactor.io/repository/github/knezzz/randomcolor/badge/master)](https://www.codefactor.io/repository/github/knezzz/randomcolor/overview/master)

Pub link: https://pub.dartlang.org/packages/random_color

This library will generate random colors that are visually pleasing and can be customized by passing custom hue, saturation and brightness range.
This allows it to generate just specific color with slight differences.

###### Blue colors:

<img src="https://lh3.googleusercontent.com/gzYVhOKSoOzObIV36RjdEH3SOWHXA9_VekOgazq8DKVBqXD_kq9tmLxFpt3llqTkSwzfLAJl16MaREDrY7c9slEMrBUptHLDOVeUV1ttKOGJn3k1HuSjolPwv2TS1j7Llp3XA96PvnaM-6Wwpwgc5awOtn-64k5m_5xV9HhZMxbpCQscSagPJ0yLN5Bt-xrDBNhAYgJafH6jgxzYpRKV6leLmvaTHQ6MUJTjQXOs69ZNaaayQAgelrb8JFNp2LIVRz6Lv7MUhAOIvr2Km1-SWKW6_EolCL9CLLRbtAAs3LAegg0HnXYElg6AC14z-ohOXrck5U9MZFE1IP7BE-xKRLD7GImQz92jMLyXw12GFBQaopGZHW8bpba95ORyqMhJbEKE1NnRsRT_Zz9RNkNrkKpGyir8tiAhz-XyYE_usPAjPNi5E236vzed5qHxBklGSa9fTo4ub7t13dnrB8pe65PAZYfBCAcJt_K4ySNzdp2NMQgtqN4CUPPUOIeuxa7Lxt8wJMeM0oL0js9GV6_wABhiCzl-IoMg5ML7OnEXX-ZC61dKDnrFmIgesu_mnSRmwkBUFd72c5Tveb28-KLZK-DQ31_N4Z2aXCbM3RB9ChXG2wGDZKgx4XFt8-AwFTbMKOk1mWD13TELkyQzH9eorbljwQQVdG--qDSlL3KP2khXNL0DIBuhGl1l=w720-h717-no" width=300>

###### Red colors:

<img src="https://lh3.googleusercontent.com/HOex8jpzv5eDKCX04nerST5HmwAGkMLe9D_FnhnxfwUDlTaboLNVmPQ2R2sdcqO_ItwdAT78vedmOt3FLgk9GVEgSd7LZYANQT7p7-P9VXcUYx6QNNQgp9axDnqENa3kD1h8vwbr8KN280QFOhssZ62Kkbc4ZBal1UzgtFbyaf1DRzmVLvqSqpaVlOSJIYsVj4stol2cvUtDzAbtKsi0btO8MenOJndIz4_efIY9gdJWjTjq4c-bj4QWwBUgyOp8fZDqmx0B-VeyT-28Z83CjO0NNV-SE-qk7Fe1Bt7XUSHlulygSpDgJhm9YalTvjBIEdwanDqlrjTN7NW_IDdEg2NI236EPIAWDd0IVN1ferff24ZBcBsRI2U8DROsDQODPiVdiXBKNPG_WDA2BTQRkoHRl-imSTD7mbsl31WBFUN7PKAvQIAh4uAnKbO2vBVUEq2chV0AWfN6_aV6L227b45-31_cpWq6mYWAS3OJWYJxEYU2Fx3Wnx8yhZNyeW2wB-uIltHh-WcZTF0Z02GbYk99HuFxP4Ikh3SYiYwlcqG2n1_1CS75IKowjY3ZBXH6xoUIuj-wAP8f2815bOllyRRwG14oj_wIPjyhvcttfKOWMFcL1Fh3kyuJsg0OF_Fuwi5yNDeT9SOWes_YohCOfpJLwNZProRxmMs1-4aiCFFx212Y4jUHfRCY=w720-h718-no" width=300>

###### Green:

<img src="https://lh3.googleusercontent.com/bs99VYp0mfYNdhMvCXeMdiTh5EQc_7uu25FkoHGOy2x2bTiysWk2eV7RED4E8ebmYgCA5X9KT1yScnLp8ayjQCYvh_N3X9shDS0bUYFYfOeCRFZap5Qk3YLPF6anRs-PQD9jGvLG3OhxmjhD3oU4LyTm5FQlnwgbIvNITYh0MN_b79W_udn7_Ow5G2kBSj1lu2r1OuCrr_rU6Ab1ILcTebu8HRi-LmoOnEMqcL2y76XPMPfXUAFGrj1mRUhQmjH0VyoH2YfI8A0ZYORqY6qO7JQ7ZL0LnIIPIxH0aoHHl6Pk2EhJyPdD5WBkWh1SQSplzGPev47bReNnNjv6CEtM_vR5DQo45d6Kc_VKVeyhp1Y9XbfJ4B_QWZvxeNu96f8XtjZ7BC3EAxU3UG4uYlnXCeem2776Mqvs6TTqnA4I1BtgzuoBMLp_oHaO5uZIlmnMJxbveBQAEhPMBb5IWRN1wypIwhewscMGSmHTfNEIBUSpPeCK-FrAHnsTjVwWOhHzx5DotgT9loeCtqiTkKyqG0cJFiTW95g_I3Oki3cD3jqalJo10JaJjr4NpYqk82xxYuRCi6XdWxQJW_R-zOh1kmaw8SD1WMFJ0RYJogOy8hvKGx6ThdmSYhrPcXAhDvH0FDJF88EZVi0lY1zJpMgzENoqNPNNLHHVAhgs4HPvHYdT5rmzIkcWsTD8=w720-h717-no" width=300>

###### Low saturation:

<img src="https://lh3.googleusercontent.com/m-UT51ts8sJFTY5QfU-lb2mxwI9MrGNixSmh4T-ue6_Y2lpw9R437cn4R-96o4VOXoLHn0Vc62PS4mgJC9oySoMGjUYNKE0qe6pBQKPjQmS7JFuliM5r-DNDqS6VN4nbdWHoItY7DM844Sb1k3SGFloeDXTRALWlFdp9m0LWImM_7K4Kcv1ZjjIiSDLX7Sz6kuKYtP6E_6-qyy7DUQ3zTIAecVxudDUL08pDAvU4Qs8_9GsgLwvIQ8k7MDFAVobIsNZpGIpKI_Px2FNqu32tlZFRycyTkXNUufgBN_a78REC5GzEGm8V3xEwDA4utAk4F2LAPWNmcmETnx2M5chefenAYa86yv2Oi7Cl4cOBR-TzQeTXgO-1Yo-Gdk1L0_KKQRRiJ3ABEko1DrkL9zEnNRaO27_G8F2Xe4KMyPE0V7l91z1aR-dPmyOFpnKPNFjfyleekBAOFx3U5mcnNlhoPH2dJybFXPC1eMBoDDUM8kJvtjrN3iY6O5wha8avrO9jhNQUtfHXihd-6mYEtKfyGX3a8AVAXf_Zwx2Em2gsbyyCzObBxBcKeiqjrvxj5B5YTMpYx5KvM8jrthg9eJ4feqAmF6k8XCGHYR1OEqueSkB-sDCyIeGqhLH4Ox2ybndPU4hR3eXEliqZlYD8GQxYKaQegNxT53R2ozf7btV3ncVfRx2pX3eh1cGO=w720-h716-no" width=300>

###### High saturation:

<img src="https://lh3.googleusercontent.com/ve62lifZ4nltAV1mXpAxl6aCn94JOA_1EdBUx5gN7ktb0LqL03M2eu_aFD4HFhzEI5ti2GGFYiwo5f_GybgX_ABGQDbF8Qefc5YlY_fOkeLBj-dOyI2xAkTmd0tKmjkayQU0Y1yhPHMue_HqdjQ_voCyH7hjRJLzUvQUzA_tl_KDPmL3Rzt8NOzF73wH9vO_a1hCip0aniUbEZyA4600jR3Xak8do0UsXxD2hmL7BgKpmatV5MMQL0MajnqcOZhLmBW8v9MioYjvXo_YV5T7olJnWM6rDXUTwHR4hPYPfbOgGFuxKeTcO3q3_yK5ez4bqzhjkoDNAEgexfMppKbB2Pi4zpD4LwADLyVhw6MSvcYuHc3SXxCuLS2yMHIs2z75K8m5dcK3iFI0pYm943N2-8JwmhYgGYXKBQ0bu6bcZ9mRURULjDrRFYSQLSoEJ-zN9NLoQXxTfSWWNAfN_1JKvbfys1XNG2EFjH9X5x53fs3ftqbxBnK7FylF9B8bZ9q0hrtLRkiOHbJKZi1HHpXiJYJDUgh5ENjnV4-ZOHtb-jcwhqw04QHROSNWzHYlKPQlWcbtTFyKoD5i3z0_HnYB4Y7VRQpEY6ZBjTwhQm61swCaW7DTAjLczfEkDs0xPuXuw986Mw98mP0DWQvNphIKaerKjFLSftOqhQZ3hkMW5GxjyIcT7A5yBqy5=w720-h715-no" width=300>

###### Light colors:

<img src="https://lh3.googleusercontent.com/DT8OZFF_Y22biSQ8LIFQKb2Oh7dl2OkqG1QE-VjDOEjJb6n1f5yOaCTuLWTu0SDK5zHekCHeigMdhEYuftYvME2hYpNRTnOn2OX_RYAl9P9jB_mF_JOIHtUS-tFxyvlsIzrA1RPOY0Sl2Jf3ruVoaKqMtHAOc-AKqbg9gJQTRt4dZZSEEY3roC_Lp9QHcEFlGF6zKTkCCNfp3N-MdKbaLFIJRH2jitd4FktQA6Syh66erFQwjMmTBwKH0wEhWKfcElze8itxnoafCoXnEGH3Nq2tpoNXHIf-FzPQHVmxYNrUbf1y0lhN8Mk0lyQXhshHD6D0ZGk5TJig1WW3CL8IUIMefXr-GqSgtwx0cUZcN7CN5borLuUSsUaKm3FYMI7I8bH4Bo-PdYZtFcPB_rG0OAZ7N19m2HaXh2DvmT-VwmS8dEMnHnW4pCCd_kf0hJZazXup_S3zVpJIi-FDkFC8v9DoCeA7GmYapRweApn3nyXnaXAJrwEDJvq5f167dSNIHXN8tW0xzMJmkX_KMq6SS2N7teHE-6BamqhuMpNiS1IYvEvEQcQgPZ3UHWaMf8SwK4CHf7ZAKyAl2pfiZJED1PN8ngnXi2fsj8t2EYQsL8ZmiD78cFD0DhZU73sjy337ankIyU-eqj4DIsUi3v111jr3xSBfPPCQ3ljdOsoAjMDOpGUH3QydjlI3=w720-h715-no" width=300>

###### Dark colors:

<img src="https://lh3.googleusercontent.com/VNTYNLKXu7gDXrz5VzISxKxeYoRhBOOUlTgS4MO-A6mKbRVT3pKaTZQK-s2PK-KJtBX0AlrP3EHhFOlnY0D_L_lRtNW-RxNcGuQYSZ7ioFMdeQEu0OtzUH5Nb55kYvkarSS7XLODcFda1srDqejuj4Rmir6HSdXtNG-B_tt2KKUPHxoFtn6gRmzn9f5pORsjEx9TsML-zs8nBeVgz7qJ5WiynbQTm_l8QOGPtHZcZrTIBjmZ1dljSlNxY9kEvLJZvMZmzZ7hVOPNJTToh1d4WbhWCZ0xeVhBshpTqRSi7ms1lbHulLh6UuidA7LiTN58wiG_XIho25EqBbqRUg46WT4God9yFMVrkd21xg372kjLb2AII2IsVKGwMBSTAZ78ZNJ4Vwj2Ivku4iwg_CjitAwdl0oLQaCtLbkjcuzcZtIJGTDmbVsrPpTV9lhJbsk4s8pEjfpldV4yAWIOo-Xatt9Jgr5c6J0imVaNTlDZqwO2Jqh7fg6hq-K4Ga7XtZG8Ct8d7z1pppNyKJYhnh6IypVAYj1ZXOpTmkad9lli0SogO41cLbbC4_VpgkMivRcboc0T77A17sYAEsb5BssVXgENVO1O4aNb0HdpicKGrCTnygtim4AKmFCd2k-wYlSgWuswcZFP4jfIonwv-kGAsCXNF3gZ15q9MbtbSuoveEyoTYkLdXEws0QO=w720-h714-no" width=300>

### Usage

##### Getting random color:
random color function can be called with parameters colorHue, colorSaturation and colorBrightness
allowing for customization of colors the generator will pass.
```dart
import 'package:random_color/random_color.dart';

RandomColor _randomColor = RandomColor();

Color _color = _randomColor.randomColor();
```

##### Getting color name:
You can get color name from hex string (without # and alpha hex values) or from Color object
```dart
  Color _color = Color('FF0000'.toStringRadix(16));
  MyColor _myColor = getColorNameFromColor(_color);

  // Prints: 'Red'
  print(_myColor.getName);
```

##### Get just red colors:
Random color can generate just specific color type, as long hue, saturation and brightness are not
too limiting this will still give rich result
```dart
import 'package:random_color/random_color.dart';

RandomColor _randomColor = RandomColor();

Color _color = _randomColor.randomColor(colorHue: ColorHue.red);
```

##### Get just red and blue colors:
You can combine multiple hue ranges to choose from
```dart
import 'package:random_color/random_color.dart';

RandomColor _randomColor = RandomColor();

Color _color = _randomColor.randomColor(
  colorHue: ColorHue.multiple([ColorHue.red, ColorHue.blue])
);
```

##### Get highly saturated colors:
```dart
import 'package:random_color/random_color.dart';

RandomColor _randomColor = RandomColor();

Color _color = _randomColor.randomColor(
  colorSaturation: ColorSaturation.highSaturation
);
```

##### Get light colors:
```dart
import 'package:random_color/random_color.dart';

RandomColor _randomColor = RandomColor();

Color _color = _randomColor.randomColor(
  colorBrightness: ColorBrightness.light
);
```

### Tests

If you want to run the tests for this project under Flutter Dart

```
flutter pub pub run test
```



CHANGELOG.md
## [1.0.6-nullsafety] - 2020.04.04

* Prepare for null-safety
* Fix for #9 (The method 'returnHue' was called on null)


## [1.0.5] - 2020.04.04

* BREAKING CHANGE: Range + (addition) will no longer mutate the object but return new instance of the object (range is const now)
* Added == and hashcode to color saturation, brightness and hue.

## [1.0.4] - 2019.08.16

* BREAKING CHANGE: Range + (addition) will no longer mutate the object but return new instance of the object (range is const now)
* Added == and hashcode to color saturation, brightness and hue.

## [1.0.3] - 2018.10.20

* Update example app
* Set debug value default to false

## [1.0.2] - 2018.10.20

* Update description

## [1.0.1] - 2018.10.17

* Remove test dependency
* Add description
* Add randomColors function that returns lost of random colors
* Change named parameter from ColorHue.custom() to required unnamed parameter
* Change named parameter from ColorSaturation.custom() to required unnamed parameter
* Change named parameter from ColorBrightness.custom() to required unnamed parameter

## [1.0.0] - TODO: 2018.10.17

* Initial release


test/random_color_test.dart
import 'dart:math';

import 'package:flutter_test/flutter_test.dart';
import 'package:random_color/random_color.dart';

void main() {
  test('Static range test', () {
    final _range = Range.staticValue(10);

    expect(_range.randomWithin(Random()), 10);
    expect(_range.randomWithin(Random()), 10);
    expect(_range.randomWithin(Random()), 10);
  });

  test('Zero range test', () {
    final _range = Range.zero();

    expect(_range.randomWithin(Random()), 0);
    expect(_range.randomWithin(Random()), 0);
    expect(_range.randomWithin(Random()), 0);
  });

  test('Range max test', () {
    final _range = Range(0, 9);

    expect(_range.randomWithin(Random()), lessThan(10));
    expect(_range.randomWithin(Random()), lessThan(10));
    expect(_range.randomWithin(Random()), lessThan(10));
    expect(_range.randomWithin(Random()), lessThan(10));
    expect(_range.randomWithin(Random()), lessThan(10));
    expect(_range.randomWithin(Random()), lessThan(10));
    expect(_range.randomWithin(Random()), lessThan(10));
    expect(_range.randomWithin(Random()), lessThan(10));
    expect(_range.randomWithin(Random()), lessThan(10));
  });

  test('Range min test', () {
    final _range = Range(91, 100);

    expect(_range.randomWithin(Random()), greaterThan(90));
    expect(_range.randomWithin(Random()), greaterThan(90));
    expect(_range.randomWithin(Random()), greaterThan(90));
    expect(_range.randomWithin(Random()), greaterThan(90));
    expect(_range.randomWithin(Random()), greaterThan(90));
    expect(_range.randomWithin(Random()), greaterThan(90));
    expect(_range.randomWithin(Random()), greaterThan(90));
    expect(_range.randomWithin(Random()), greaterThan(90));
    expect(_range.randomWithin(Random()), greaterThan(90));
  });

  test('Color name test', () {
    expect(getColorNameFromString('FF0000').getName.toLowerCase(), 'red');
    expect(getColorNameFromString('00FF00').getName.toLowerCase(), 'green');
    expect(getColorNameFromString('0000FF').getName.toLowerCase(), 'blue');
  });

  test('Color generation', () {
    final rc = RandomColor();

    final _redColor = rc.randomColor(
        colorHue: ColorHue.custom(Range.zero()),
        colorSaturation: ColorSaturation.custom(Range.staticValue(100)),
        colorBrightness: ColorBrightness.custom(Range.staticValue(50)),
        debug: false);

    final _greenColor = rc.randomColor(
        colorHue: ColorHue.custom(Range.staticValue(120)),
        colorSaturation: ColorSaturation.custom(Range.staticValue(100)),
        colorBrightness: ColorBrightness.custom(Range.staticValue(50)),
        debug: false);

    final _blueColor = rc.randomColor(
        colorHue: ColorHue.custom(Range.staticValue(240)),
        colorSaturation: ColorSaturation.custom(Range.staticValue(100)),
        colorBrightness: ColorBrightness.custom(Range.staticValue(50)),
        debug: false);

    expect(getColorNameFromColor(_redColor).getName.toLowerCase(), 'red');
    expect(getColorNameFromColor(_greenColor).getName.toLowerCase(), 'green');
    expect(getColorNameFromColor(_blueColor).getName.toLowerCase(), 'blue');
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:random_color/random_color.dart';

void main() => runApp(MaterialApp(home: MyApp()));

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  List<Color> generatedColors = <Color>[];
  final List<ColorHue> _hueType = <ColorHue>[
    ColorHue.green,
    ColorHue.red,
    ColorHue.pink,
    ColorHue.purple,
    ColorHue.blue,
    ColorHue.yellow,
    ColorHue.orange
  ];
  ColorBrightness _colorLuminosity = ColorBrightness.random;
  ColorSaturation _colorSaturation = ColorSaturation.random;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Example of Random colors library'),
      ),
      body: Container(child: _showColors()),
      floatingActionButton: Column(
        mainAxisAlignment: MainAxisAlignment.end,
        children: <Widget>[
          Padding(
            padding: const EdgeInsets.symmetric(vertical: 8.0),
            child: FloatingActionButton(
              tooltip: 'RefreshColors',
              child: const Icon(Icons.refresh),
              onPressed: _updateColor,
            ),
          ),
          Padding(
            padding: const EdgeInsets.symmetric(vertical: 8.0),
            child: FloatingActionButton(
              tooltip: 'Filter',
              child: const Icon(Icons.filter_list),
              onPressed: _showFilterDialog,
            ),
          ),
        ],
      ),
    );
  }

  Widget _showColors() {
    return GridView.builder(
      gridDelegate:
          const SliverGridDelegateWithFixedCrossAxisCount(crossAxisCount: 3),
      shrinkWrap: true,
      itemBuilder: (BuildContext context, int index) {
        Color _color;

        if (generatedColors.length > index) {
          _color = generatedColors[index];
        } else {
          _color = RandomColor().randomColor(
              colorHue: ColorHue.multiple(colorHues: _hueType),
              colorSaturation: _colorSaturation,
              colorBrightness: _colorLuminosity);

          generatedColors.add(_color);
        }

        Color getTextColor() {
          if (_color.computeLuminance() > 0.3) {
            return Colors.black;
          } else {
            return Colors.white;
          }
        }

        return InkWell(
          onTap: () {
            showDialog<void>(
                context: context,
                builder: (BuildContext context) {
                  final MaterialColor _mc = RandomColor().randomMaterialColor(
                      colorHue: ColorHue.custom(Range.staticValue(
                          HSLColor.fromColor(_color).hue.toInt())),
                      colorSaturation: _colorSaturation);

                  return Dialog(
                    child: Column(
                      children: <Widget>[
                        const Text('Material color'),
                        Column(
                          children: <Widget>[
                            Container(
                              height: 50.0,
                              color: _mc.shade50,
                              child: const Center(
                                child: Text('50'),
                              ),
                            ),
                            Container(
                              height: 50.0,
                              color: _mc.shade100,
                              child: const Center(
                                child: Text('100'),
                              ),
                            ),
                            Container(
                              height: 50.0,
                              color: _mc.shade200,
                              child: const Center(
                                child: Text('200'),
                              ),
                            ),
                            Container(
                              height: 50.0,
                              color: _mc.shade300,
                              child: const Center(
                                child: Text('300'),
                              ),
                            ),
                            Container(
                              height: 50.0,
                              color: _mc.shade400,
                              child: const Center(
                                child: Text('400'),
                              ),
                            ),
                            Container(
                              height: 80.0,
                              color: _mc.shade500,
                              child: const Center(
                                child: Text('500 - Base'),
                              ),
                            ),
                            Container(
                              height: 50.0,
                              color: _mc.shade600,
                              child: const Center(
                                child: Text('600'),
                              ),
                            ),
                            Container(
                              height: 50.0,
                              color: _mc.shade700,
                              child: const Center(
                                child: Text('700'),
                              ),
                            ),
                            Container(
                              height: 50.0,
                              color: _mc.shade800,
                              child: const Center(
                                child: Text('800'),
                              ),
                            ),
                            Container(
                              height: 50.0,
                              color: _mc.shade900,
                              child: const Center(
                                child: Text('900'),
                              ),
                            )
                          ],
                        )
                      ],
                    ),
                  );
                });
          },
          child: Card(
            color: _color,
            child: Container(
              margin: const EdgeInsets.all(12.0),
              child: Column(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: <Widget>[
                  Container(
                    alignment: Alignment.centerLeft,
                    child: Text(
                      getColorNameFromColor(_color).getName,
                      style: Theme.of(context)
                          .textTheme
                          .headline6
                          ?.copyWith(fontSize: 13.0, color: getTextColor()),
                    ),
                  ),
                  Container(
                    alignment: Alignment.centerRight,
                    child: Text(
                      '#${_color.value.toRadixString(16).toUpperCase()}',
                      style: Theme.of(context).textTheme.caption?.copyWith(
                          color: getTextColor(),
                          fontSize: 16.0,
                          fontWeight: FontWeight.w300),
                    ),
                  ),
                ],
              ),
            ),
          ),
        );
      },
    );
  }

  void _showFilterDialog() async {
    await showDialog<Null>(
        context: context,
        builder: (BuildContext context) {
          return FilterDialog(
            hueType: _hueType,
            colorLuminosity: _colorLuminosity,
            colorSaturation: _colorSaturation,
            hueTypeChange: (List<ColorHue> hues) {
              _hueType.clear();
              _hueType.addAll(hues);
            },
            luminosityTypeChange: (ColorBrightness cb) => _colorLuminosity = cb,
            saturationTypeChange: (ColorSaturation cs) => _colorSaturation = cs,
          );
        }).catchError(print);

    _updateColor();
  }

  void _updateColor() {
    setState(() {
      generatedColors.clear();
    });
  }
}

typedef HueTypeChange = void Function(List<ColorHue> colorHues);
typedef SaturationTypeChange = void Function(ColorSaturation colorSaturation);
typedef LuminosityTypeChange = void Function(ColorBrightness colorBrightness);

class FilterDialog extends StatefulWidget {
  const FilterDialog({
    Key? key,
    required this.hueType,
    required this.colorSaturation,
    required this.colorLuminosity,
    required this.hueTypeChange,
    required this.saturationTypeChange,
    required this.luminosityTypeChange,
  }) : super(key: key);

  final List<ColorHue> hueType;
  final ColorBrightness colorLuminosity;
  final ColorSaturation colorSaturation;

  final HueTypeChange hueTypeChange;
  final SaturationTypeChange saturationTypeChange;
  final LuminosityTypeChange luminosityTypeChange;

  @override
  _FilterDialogState createState() => _FilterDialogState();
}

class _FilterDialogState extends State<FilterDialog> {
  final List<ColorHue> _hueType = <ColorHue>[];
  late ColorBrightness _colorLuminosity;
  late ColorSaturation _colorSaturation;

  @override
  void initState() {
    super.initState();

    _hueType.addAll(widget.hueType);
    _colorLuminosity = widget.colorLuminosity;
    _colorSaturation = widget.colorSaturation;
  }

  @override
  Widget build(BuildContext context) {
    return SimpleDialog(children: <Widget>[
      Padding(
        padding: const EdgeInsets.all(8.0),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: <Widget>[
            Container(
              color: Colors.white,
              child: const Text('Color hue: '),
            ),
            Container(
              height: 175.0,
              width: MediaQuery.of(context).size.width * 0.5,
              child: ListView.builder(
                shrinkWrap: true,
                itemCount: ColorHue.values.length,
                itemBuilder: (BuildContext context, int index) {
                  final ColorHue? _hue = _hueType.firstWhere((ColorHue hue) =>
                      hue.type == ColorHue.values[index].type);
                  final Color _color = RandomColor(8).randomColor(
                      colorHue: ColorHue.values[index],
                      colorSaturation: _colorSaturation,
                      colorBrightness: _colorLuminosity);

                  return Container(
                    height: 50.0,
                    child: Flex(
                      direction: Axis.horizontal,
                      children: <Widget>[
                        Container(
                          height: 50.0,
                          width: 12.0,
                          color: _color,
                        ),
                        Expanded(
                          child: Container(
                            color: _color.withOpacity(0.1),
                            padding: const EdgeInsets.all(8.0),
                            child: Row(
                              mainAxisAlignment: MainAxisAlignment.spaceBetween,
                              children: <Widget>[
                                Text(ColorHue.values[index].toString()),
                                Checkbox(
                                  value: _hue != null,
                                  onChanged: (bool? value) {
                                    if (_hue != null) {
                                      _removeColorHue(_hue);
                                    } else {
                                      _addColorHue(ColorHue.values[index]);
                                    }

                                    _hueType.removeWhere(
                                        (ColorHue hue) => hue == null);
                                    widget.hueTypeChange(_hueType);
                                  },
                                )
                              ],
                            ),
                          ),
                        ),
                      ],
                    ),
                  );
                },
              ),
            ),
          ],
        ),
      ),
      Padding(
        padding: const EdgeInsets.all(8.0),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: <Widget>[
            Container(
              color: Colors.white,
              child: const Text('Color brightness: '),
            ),
            DropdownButton<int>(
              value: _colorLuminosity.type,
              onChanged: (int? luminosity) {
                setState(() =>
                    _colorLuminosity = ColorBrightness.values[luminosity ?? 0]);

                widget.luminosityTypeChange(_colorLuminosity);
              },
              items: ColorBrightness.values
                  .map((ColorBrightness l) => DropdownMenuItem<int>(
                      child: Container(child: Text(l.toString())),
                      value: l.type))
                  .toList(),
            )
          ],
        ),
      ),
      Padding(
        padding: const EdgeInsets.all(8.0),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: <Widget>[
            Container(
              color: Colors.white,
              child: const Text('Color saturation: '),
            ),
            DropdownButton<int>(
              value: _colorSaturation.type,
              onChanged: (int? saturation) {
                setState(() =>
                    _colorSaturation = ColorSaturation.values[saturation ?? 0]);

                widget.saturationTypeChange(_colorSaturation);
              },
              items: ColorSaturation.values
                  .map((ColorSaturation cf) => DropdownMenuItem<int>(
                      child: Container(child: Text(cf.toString())),
                      value: cf.type))
                  .toList(),
            )
          ],
        ),
      ),
    ]);
  }

  void _addColorHue(ColorHue c) {
    setState(() {
      _hueType.add(c);
    });
  }

  void _removeColorHue(ColorHue c) {
    setState(() {
      _hueType.remove(c);

      if (_hueType.isEmpty) {
        _hueType.add(ColorHue.random);
      }
    });
  }
}


lib/random_color/color_info.dart
part of random_color;

@immutable
class ColorBrightness {
  const ColorBrightness.custom(Range brightnessRange)
      : _brightness = brightnessRange,
        type = -1;

  const ColorBrightness._(this._brightness, this.type);

  static const int minBrightness = 16;
  static const int maxBrightness = 84;

  static const ColorBrightness dark =
      ColorBrightness._(Range(minBrightness, minBrightness + 30), 3);
  static const ColorBrightness light = ColorBrightness._(
      Range((maxBrightness + minBrightness) ~/ 2, maxBrightness), 1);

  static const ColorBrightness primary =
      ColorBrightness._(Range(minBrightness + 20, maxBrightness - 20), 2);
  static const ColorBrightness random =
      ColorBrightness._(Range(minBrightness, maxBrightness), 5);

  static const List<ColorBrightness> values = <ColorBrightness>[
    veryLight,
    light,
    primary,
    dark,
    veryDark,
    random
  ];
  static const ColorBrightness veryDark =
      ColorBrightness._(Range(minBrightness ~/ 2, minBrightness + 30), 4);
  static const ColorBrightness veryLight = ColorBrightness._(
      Range((maxBrightness + minBrightness) ~/ 2,
          maxBrightness + (minBrightness ~/ 2)),
      0);
  final Range _brightness;
  final int type;

  int returnBrightness(Random random) {
    return _brightness.randomWithin(random);
  }

  @override
  String toString() {
    switch (type) {
      case 0:
        return 'very light';
      case 1:
        return 'light';
      case 2:
        return 'primary';
      case 3:
        return 'dark';
      case 4:
        return 'very dark';
      case 5:
        return 'random';
    }

    return 'custom';
  }

  static ColorBrightness multiple(
      {required List<ColorBrightness> colorBrightnessList, Random? random}) {
    colorBrightnessList.shuffle(random);
    return colorBrightnessList.first;
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ColorBrightness &&
          runtimeType == other.runtimeType &&
          _brightness == other._brightness &&
          type == other.type;

  @override
  int get hashCode => _brightness.hashCode ^ type.hashCode;
}

@immutable
class ColorHue {
  const ColorHue.custom(Range hueRange)
      : _hue = hueRange,
        type = -1;
  const ColorHue._(this._hue, this.type);

  static const ColorHue blue = ColorHue._(Range(180, 240), 4);

  static const ColorHue green = ColorHue._(Range(60, 180), 3);

  static const ColorHue orange = ColorHue._(Range(10, 40), 1);

  static const ColorHue pink = ColorHue._(Range(315, 355), 6);
  static const ColorHue purple = ColorHue._(Range(240, 315), 5);

  static const ColorHue random = ColorHue._(Range(0, 360), 7);
  static const ColorHue red = ColorHue._(Range(-5, 10), 0);
  static const List<ColorHue> values = <ColorHue>[
    red,
    orange,
    yellow,
    green,
    blue,
    purple,
    pink,
    random
  ];
  static const ColorHue yellow = ColorHue._(Range(40, 60), 2);
  final Range _hue;
  final int type;

  int returnHue(Random random) {
    var _h = _hue.randomWithin(random);

    if (_h < 0) {
      _h = 360 + _h;
    }

    return _h;
  }

  @override
  String toString() {
    switch (type) {
      case 0:
        return 'red';
      case 1:
        return 'orange';
      case 2:
        return 'yellow';
      case 3:
        return 'green';
      case 4:
        return 'blue';
      case 5:
        return 'purple';
      case 6:
        return 'pink';
      case 7:
        return 'random';
    }

    return 'custom';
  }

  static ColorHue multiple(
      {required List<ColorHue> colorHues, Random? random}) {
    colorHues.shuffle(random);
    return colorHues.first;
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ColorHue &&
          runtimeType == other.runtimeType &&
          _hue == other._hue &&
          type == other.type;

  @override
  int get hashCode => _hue.hashCode ^ type.hashCode;
}

@immutable
class ColorSaturation {
  const ColorSaturation.custom(Range saturationRange)
      : _saturation = saturationRange,
        type = -1;

  const ColorSaturation._(this._saturation, this.type);

  static const ColorSaturation highSaturation =
      ColorSaturation._(Range(80, 100), 2);

  static const ColorSaturation lowSaturation =
      ColorSaturation._(Range(0, 40), 0);

  static const ColorSaturation mediumSaturation =
      ColorSaturation._(Range(40, 80), 1);
  static const ColorSaturation monochrome = ColorSaturation._(Range.zero(), 4);

  static const ColorSaturation random = ColorSaturation._(Range(20, 100), 3);
  static const List<ColorSaturation> values = <ColorSaturation>[
    lowSaturation,
    mediumSaturation,
    highSaturation,
    random,
    monochrome
  ];
  final Range _saturation;
  final int type;

  int returnSaturation(Random random) {
    return _saturation.randomWithin(random);
  }

  @override
  String toString() {
    switch (type) {
      case 0:
        return 'low saturation';
      case 1:
        return 'medium saturation';
      case 2:
        return 'high saturation';
      case 3:
        return 'random';
      case 4:
        return 'monochrome';
    }

    return 'custom';
  }

  static ColorSaturation multiple(
      {required List<ColorSaturation> colorSaturations, Random? random}) {
    colorSaturations.shuffle(random);
    return colorSaturations.first;
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ColorSaturation &&
          runtimeType == other.runtimeType &&
          _saturation == other._saturation &&
          type == other.type;

  @override
  int get hashCode => _saturation.hashCode ^ type.hashCode;
}

@immutable
class Range {
  const Range(this.start, this.end);

  const Range.staticValue(int value)
      : start = value,
        end = value;
  const Range.zero()
      : start = 0,
        end = 0;

  final int start;
  final int end;

  Range operator +(Range range) {
    return Range((start + range.start) ~/ 2, end);
  }

  bool contain(int value) {
    return value >= start && value <= end;
  }

  int randomWithin(Random random) {
    return (start + random.nextDouble() * (end - start)).round();
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Range &&
          runtimeType == other.runtimeType &&
          start == other.start &&
          end == other.end;

  @override
  int get hashCode => start.hashCode ^ end.hashCode;
}
